test/e2e/network/network_policy.go

```diff
@@ -852,20 +852,75 @@ var _ = SIGDescribe("NetworkPolicy [LinuxOnly]", func() {
 			defer cleanupNetworkPolicy(f, policy)
 
 			ginkgo.By(fmt.Sprintf("Creating client pod %s that should not be able to connect to %s.", "client-a", service.Name))
-			podClient := createNetworkClientPod(f, f.Namespace, "client-a", service, allowedPort)
+			// Specify RestartPolicy to OnFailure so we can check the client pod fails in the beginning and succeeds
+			// after updating its label, otherwise it would not restart after the first failure.
+			podClient := createNetworkClientPodWithRestartPolicy(f, f.Namespace, "client-a", service, allowedPort, v1.RestartPolicyOnFailure)
 			defer func() {
 				ginkgo.By(fmt.Sprintf("Cleaning up the pod %s", podClient.Name))
 				if err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Delete(podClient.Name, nil); err != nil {
 					framework.Failf("unable to cleanup pod %v: %v", podClient.Name, err)
 				}
 			}()
-			checkNoConnectivity(f, f.Namespace, podClient, service)
+			// Check Container exit code as restartable Pod's Phase will be Running even when container fails.
+			err = e2epod.WaitForPodCondition(f.ClientSet, f.Namespace.Name, podClient.Name, "terminated", framework.PodStartTimeout, func(pod *v1.Pod) (bool, error) {
+				statuses := pod.Status.ContainerStatuses
+				if len(statuses) == 0 || statuses[0].State.Terminated == nil {
+					return false, nil
+				}
+				// We expect an error here since it's a cannot connect test. Return error if exit code is 0.
+				if statuses[0].State.Terminated.ExitCode == 0 {
+					return true, fmt.Errorf("pod %q terminated without error", pod.Name)
+				}
+				return true, nil
+			})
+			if err != nil {
+				pods, policies, logs := collectPodsAndNetworkPolicies(f, podClient)
+				framework.Failf("Pod %s should not be able to connect to service %s, but was able to connect.\nPod logs:\n%s\n\n Current NetworkPolicies:\n\t%v\n\n Pods:\n\t%v\n\n", podClient.Name, service.Name, logs, policies.Items, pods)
+
+				// Dump debug information for the test namespace.
+				framework.DumpDebugInfo(f.ClientSet, f.Namespace.Name)
+			}
 
 			ginkgo.By(fmt.Sprintf("Updating client pod %s that should successfully connect to %s.", podClient.Name, service.Name))
-			podClient = updateNetworkClientPodLabel(f, f.Namespace, podClient.Name, "replace", "/metadata/labels", map[string]string{})
+			podClient = updatePodLabel(f, f.Namespace, podClient.Name, "replace", "/metadata/labels", map[string]string{})
 			checkConnectivity(f, f.Namespace, podClient, service)
```

Comment:
If checkConnectivity fails, who will clean up the pod as it would keep restarting on failure?

Reply:
`checkConnectivity` has timeout (3 min):
https://github.com/kubernetes/kubernetes/blob/ae95a4bfcac12bf6b19e2d5acf6404359b1a8c3b/test/e2e/framework/pod/wait.go#L356-L357
The test will fail if timeout occurs and the Pod will be cleaned up by the deferred delete function.